src/SFML/Window/Win32/JoystickImpl.cpp

```diff
@@ -44,1115 +44,724 @@
-    indexString << index + 1;
-
-    subkey = TEXT("Joystick");
-    subkey += indexString.str();
-    subkey += REGSTR_VAL_JOYOEMNAME;
-
-    std::array<TCHAR, 256> keyData{};
-    DWORD                  keyDataSize = sizeof(keyData);
-
-    result = RegQueryValueEx(currentKey, subkey.c_str(), nullptr, nullptr, reinterpret_cast<LPBYTE>(keyData.data()), &keyDataSize);
-    RegCloseKey(currentKey);
-
-    if (result != ERROR_SUCCESS)
-    {
-        sf::err() << "Unable to query registry key for joystick at index " << index << ": "
-                  << sf::priv::getErrorString(static_cast<DWORD>(result)) << std::endl;
-        return joystickDescription;
-    }
-
-    subkey = REGSTR_PATH_JOYOEM;
-    subkey += TEXT('\\');
-    subkey.append(keyData.data(), keyDataSize / sizeof(TCHAR));
-
-    result = RegOpenKeyEx(rootKey, subkey.c_str(), 0, KEY_READ, &currentKey);
-
-    if (result != ERROR_SUCCESS)
-    {
-        sf::err() << "Unable to open registry key for joystick at index " << index << ": "
-                  << sf::priv::getErrorString(static_cast<DWORD>(result)) << std::endl;
-        return joystickDescription;
-    }
-
-    keyDataSize = sizeof(keyData);
-
-    result = RegQueryValueEx(currentKey, REGSTR_VAL_JOYOEMNAME, nullptr, nullptr, reinterpret_cast<LPBYTE>(keyData.data()), &keyDataSize);
-    RegCloseKey(currentKey);
-
-    if (result != ERROR_SUCCESS)
-    {
-        sf::err() << "Unable to query name for joystick at index " << index << ": "
-                  << sf::priv::getErrorString(static_cast<DWORD>(result)) << std::endl;
-        return joystickDescription;
-    }
-
-    keyData.back() = TEXT('\0'); // Ensure null terminator in case the data is too long.
-    return keyData.data();
 }
-} // namespace
 
-namespace sf::priv
-{
-////////////////////////////////////////////////////////////
-void JoystickImpl::initialize()
+bool extractVidPid(const std::wstring& devicePath, USHORT& vid, USHORT& pid, bool& isXInput)
 {
-    // Try to initialize DirectInput
-    initializeDInput();
-
-    if (!directInput)
-        err() << "DirectInput not available, falling back to Windows joystick API" << std::endl;
-
-    // Perform the initial scan and populate the connection cache
-    updateConnections();
-}
-
-
-////////////////////////////////////////////////////////////
-void JoystickImpl::cleanup()
-{
-    // Clean up DirectInput
-    cleanupDInput();
-}
-
-
-////////////////////////////////////////////////////////////
-bool JoystickImpl::isConnected(unsigned int index)
-{
-    if (directInput)
-        return isConnectedDInput(index);
-
-    ConnectionCache&   cache                  = connectionCache[index];
-    constexpr sf::Time connectionRefreshDelay = sf::milliseconds(500);
-    if (!lazyUpdates && cache.timer.getElapsedTime() > connectionRefreshDelay)
+    // Find VID
+    std::size_t vidPos = devicePath.find(L"VID_");
+    if (vidPos == std::wstring::npos)
+        return false;
+    vidPos += 4;
```

Comment:
Where does this `4` come from? Can we replace some of these magic numbers with more expressive expressions or a named constant if that's the right solution? This goes for some other magic numbers in this file. It's hard for me to figure out what the code is trying to do or how I can change it without breaking an internal invariant.